android/libraries/rib-base/src/main/java/com/badoo/ribs/core/routing/RoutingSource.kt

```diff
@@ -0,0 +1,190 @@
+        }
+
+        private val combined = Observable.combineLatest(
+            first,
+            second,
+            BiFunction<RoutingHistory<C>, RoutingHistory<C>, RoutingHistory<C>> { source1, source2 ->
+                CombinedHistory(source1, source2)
+            }
+        )
+
+        override fun subscribe(observer: Observer<in RoutingHistory<C>>) {
+            combined.subscribe(observer)
+        }
+
+        override fun remove(identifier: Identifier) {
+            first.remove(identifier)
+            second.remove(identifier)
+        }
+
+        override fun handleBackPressFirst(): Boolean =
+            first.handleBackPressFirst() || second.handleBackPressFirst()
+
+        override fun handleBackPressFallback(): Boolean =
+            first.handleBackPressFallback() || second.handleBackPressFallback()
+    }
+
+    // TODO extract
+    class Permanent<C : Parcelable>(permanents: Iterable<C>) : RoutingSource<C> {
+
+        companion object {
+            fun <C : Parcelable> permanent(permanents: Iterable<C>) = Permanent(permanents)
+
+            fun <C : Parcelable> permanent(vararg permanents: C) = Permanent(permanents.toSet())
+        }
+
+        private val routingElements =
+            permanents.mapIndexed { idx, configuration ->
+                RoutingHistoryElement(
+                    activation = ACTIVE,
+                    routing = Routing(
+                        configuration = configuration,
+                        identifier = Identifier("Permanent $idx")
+                    )
+            ) }
+
+        private val permanentHistory =
+            RoutingHistory.from(routingElements)
+
+        override fun remove(identifier: Identifier) {
+            // no-op -- it's permanent!
+        }
+
+        override fun subscribe(observer: Observer<in RoutingHistory<C>>) {
+            Observable
+                .just(permanentHistory)
+                .subscribe(observer)
+        }
+    }
+
+    // TODO extract
+    class Empty<C : Parcelable> : RoutingSource<C> {
+        override fun remove(identifier: Identifier) {
+            // no-op -- it's empty
+        }
+
+        override fun subscribe(observer: Observer<in RoutingHistory<C>>) =
+            Observable.empty<RoutingHistory<C>>().subscribe(observer)
+    }
+
+    // TODO extract
+    class Pool<C : Parcelable>(
+        private val allowRepeatingConfigurations: Boolean
+    ) : RoutingSource<C> {
+        private var elements: Map<Identifier, RoutingHistoryElement<C>> = emptyMap()
+        private val current: RoutingHistory<C>
+            get() = RoutingHistory.from(elements.values)
+
+        private val states: Relay<RoutingHistory<C>> = BehaviorRelay.createDefault(
+            RoutingHistory.from(current)
+        )
+
+        fun add(
+            configuration: C,
+            identifier: Identifier = Identifier(
+                "Set ${System.identityHashCode(this)} #$configuration"
```

Comment:
so, am I right that id for Identifier will be same for any added configuration, if client will not pass Identifier here

Reply:
good catch! will update